### PR TITLE
Add StreamInfo structure

### PR DESCRIPTION
**Pull request by naushir**

### CODE REVIEW

**Comment by naushir** [comment]
@davidplowman need another set of eyes to make sure I've not messed up.

File: apps/libcamera_still.cpp [hunk 1]
```diff
@@ -108,7 +107,7 @@ static void save_images(LibcameraStillApp &app, CompletedRequestPtr &payload)
 		options->framestart %= options->wrap;
 }
 
-// Some keypress/signal handling.
+// Some keypress/signal info.heightandling.
```
Comment (naushir) [inline apps/libcamera_still.cpp:111]: Oops!

File: apps/libcamera_still.cpp [hunk 2]
```diff
@@ -183,7 +182,7 @@ static void event_loop(LibcameraStillApp &app)
 		if (key == 'x' || key == 'X')
 			return;
 
-		// In viewfinder mode, simply run until the timeout. When that happens, switch to
+		// In viewfinder mode, simply run until the timeout. When that info.heightappens, switch to
 		// capture mode if an output was requested.
```
Comment (naushir) [inline apps/libcamera_still.cpp:186]: Oops

File: apps/libcamera_still.cpp [hunk 3]
```diff
@@ -198,7 +197,7 @@ static void event_loop(LibcameraStillApp &app)
 			if (timed_out || keypressed || timelapse_timed_out)
 			{
 				// Trigger a still capture unless:
-				if (!output || // we have no output file
+				if (!output || // we info.heightave no output file
 					(timed_out && options->timelapse) || // timed out in timelapse mode
```
Comment (naushir) [inline apps/libcamera_still.cpp:201]: Oops

File: core/libcamera_app.hpp [hunk 4]
```diff
@@ -104,22 +105,20 @@ class LibcameraApp
...
+	Stream *LoresStream(StreamInfo *info = nullptr) const;
 	Stream *GetMainStream() const;
 
 	std::vector<libcamera::Span<uint8_t>> Mmap(FrameBuffer *buffer) const;
 
 	void ShowPreview(CompletedRequestPtr &completed_request, Stream *stream);
 
 	void SetControls(ControlList &controls);
-	void StreamDimensions(Stream const *stream, unsigned int *w, unsigned int *h, unsigned int *stride) const;
+	void GetStreamInfo(Stream const *stream, StreamInfo *info) const;
```
Comment (davidplowman) [inline core/libcamera_app.hpp:121]: I wonder very slightly about simply making this return-by-value now, but honestly, it's so trivial I don't really mind...

File: post_processing_stages/annotate_cv_stage.cpp [hunk 5]
```diff
@@ -69,7 +69,7 @@ void AnnotateCvStage::Configure()
 	stream_ = app_->GetMainStream();
 	if (!stream_ || stream_->configuration().pixelFormat != libcamera::formats::YUV420)
 		throw std::runtime_error("AnnotateCvStage: only YUV420 format supported");
-	app_->StreamDimensions(stream_, &width_, &height_, &stride_);
+	app_->GetStreamInfo(stream_, &width_, &height_, &stride_);
```
Comment (davidplowman) [inline post_processing_stages/annotate_cv_stage.cpp:72]: I thought we'd replaced this one by GetStreamInfo(stream_, &info) now? (Maybe I misunderstood something?)

File: post_processing_stages/tf_stage.cpp [hunk 6]
```diff
@@ -67,7 +67,7 @@ void TfStage::Configure()
 	lores_stream_ = app_->LoresStream();
 	if (lores_stream_)
 	{
-		app_->StreamDimensions(lores_stream_, &lores_w_, &lores_h_, &lores_stride_);
+		app_->GetStreamInfo(lores_stream_, &lores_w_, &lores_h_, &lores_stride_);
 		if (config_->verbose)
```
Comment (naushir) [inline post_processing_stages/tf_stage.cpp:71]: Missed this one.

File: post_processing_stages/tf_stage.cpp [hunk 7]
```diff
@@ -83,7 +83,7 @@ void TfStage::Configure()
 	main_stream_ = app_->GetMainStream();
 	if (main_stream_)
 	{
-		app_->StreamDimensions(main_stream_, &main_w_, &main_h_, &main_stride_);
+		app_->GetStreamInfo(main_stream_, &main_w_, &main_h_, &main_stride_);
 		if (config_->verbose)
```
Comment (naushir) [inline post_processing_stages/tf_stage.cpp:87]: Missed this one.

File: encoder/mjpeg_encoder.cpp [hunk 8]
```diff
@@ -39,11 +39,10 @@ MjpegEncoder::~MjpegEncoder()
 		std::cerr << "MjpegEncoder closed" << std::endl;
 }
 
-void MjpegEncoder::EncodeBuffer(int fd, size_t size, void *mem, unsigned int width, unsigned int height,
-								unsigned int stride, int64_t timestamp_us)
+void MjpegEncoder::EncodeBuffer(int fd, size_t size, void *mem, StreamInfo const &info, int64_t timestamp_us)
 {
 	std::lock_guard<std::mutex> lock(encode_mutex_);
-	EncodeItem item = { mem, width, height, stride, timestamp_us, index_++ };
+	EncodeItem item = { mem, info.width, info.height, info.stride, timestamp_us, index_++ };
```
Comment (davidplowman) [inline encoder/mjpeg_encoder.cpp:45]: Might we want simply to put the whole "info" into the EncodeItem? I guess it doesn't really make any difference, it's not like we're going to do anything with the pixel format or colour space.

File: preview/drm_preview.cpp [hunk 9]
```diff
@@ -269,38 +269,38 @@ DrmPreview::~DrmPreview()
...
+void DrmPreview::makeBuffer(int fd, size_t size, StreamInfo const &info, Buffer &buffer)
 {
 	buffer.fd = fd;
 	buffer.size = size;
-	buffer.width = width;
-	buffer.height = height;
-	buffer.stride = stride;
+	buffer.width = info.width;
+	buffer.height = info.height;
+	buffer.stride = info.stride;
```
Comment (davidplowman) [inline preview/drm_preview.cpp:278]: Could put the entire info in the buffer, though again, it seems scarcely worth the trouble. TBH, looking at what's below, do we even need these values at all? Maybe a question for another time.

File: preview/egl_preview.cpp [hunk 10]
```diff
@@ -338,36 +338,36 @@ void EglPreview::makeWindow(char const *name)
...
 	}
 
 	buffer.fd = fd;
 	buffer.size = size;
-	buffer.width = width;
-	buffer.height = height;
-	buffer.stride = stride;
+	buffer.width = info.width;
+	buffer.height = info.height;
+	buffer.stride = info.stride;
```
Comment (davidplowman) [inline preview/egl_preview.cpp:356]: Same as above., though I think we probably just leave this for now!

File: core/libcamera_app.hpp (hunk 4 above)
Comment (naushir) [reply core/libcamera_app.hpp:121]: Good idea.

File: encoder/encoder.cpp [hunk 11]
```diff
@@ -12,12 +12,12 @@
...
+Encoder *Encoder::Create(VideoOptions const *options, const StreamInfo &info)
 {
 	if (strcasecmp(options->codec.c_str(), "yuv420") == 0)
 		return new NullEncoder(options);
 	else if (strcasecmp(options->codec.c_str(), "h264") == 0)
-		return new H264Encoder(options);
+		return new H264Encoder(options, info);
 	else if (strcasecmp(options->codec.c_str(), "mjpeg") == 0)
 		return new MjpegEncoder(options);
 	throw std::runtime_error("Unrecognised codec " + options->codec);
```
Comment (davidplowman) [inline encoder/encoder.cpp:23]: My sense of symmetry feels slightly uncomfortable that not all the encoders want the info parameter, but I think I can survive!  :)

**Pull request merged**